File: classifier/data/raw/ai/samples/project_euler_java/Problem092.java
// Project Euler 92: square-digit chains ending at 89 below ten million.
public class Problem092 {
    private static int next(int n) {
        int sum = 0;
        while (n != 0) { int d = n % 10; sum += d * d; n /= 10; }
        return sum;
    }
    public static void main(String[] args) {
        int[] endings = new int[568];
        endings[1] = 1; endings[89] = 89;
        for (int i = 2; i < endings.length; i++) {
            int x = i;
            while (endings[x] == 0) x = next(x);
            endings[i] = endings[x];
        }
        long[] counts = {1};
        for (int position = 0; position < 7; position++) {
            long[] expanded = new long[81 * (position + 1) + 1];
            for (int sum = 0; sum < counts.length; sum++)
                for (int digit = 0; digit <= 9; digit++) expanded[sum + digit * digit] += counts[sum];
            counts = expanded;
        }
        long answer = 0;
        for (int sum = 1; sum < counts.length; sum++) if (endings[sum] == 89) answer += counts[sum];
        System.out.println(answer); // 0..9,999,999; zero itself does not end at 89.
    }
}

File: classifier/data/raw/human/samples/project_euler_zig/problem_092.zig
const std = @import("std");

// Project Euler 92: count starting values below ten million ending at 89.
fn squareDigitSum(mut_value: usize) usize {
    var value = mut_value;
    var sum: usize = 0;
    while (value != 0) : (value /= 10) {
        const digit = value % 10;
        sum += digit * digit;
    }
    return sum;
}

pub fn main() !void {
    var fate = [_]u8{0} ** 568; // A seven-digit number has sum at most 7 * 81.
    fate[1] = 1;
    fate[89] = 89;
    var count: usize = 0;
    for (1..10_000_000) |start| {
        var value = squareDigitSum(start);
        var trail: [32]usize = undefined;
        var trail_len: usize = 0;
        while (fate[value] == 0) {
            trail[trail_len] = value;
            trail_len += 1;
            value = squareDigitSum(value);
        }
        const ending = fate[value];
        for (trail[0..trail_len]) |item| fate[item] = ending;
        if (ending == 89) count += 1;
    }
    var buffer: [32]u8 = undefined;
    try std.Io.File.writeStreamingAll(std.Io.File.stdout(), std.Io.Threaded.global_single_threaded.io(), (try std.fmt.bufPrint(&buffer, "{d}\n", .{count})));
}
